analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaIntegration.java

```diff
@@ -60,25 +89,119 @@
   @Override public void identify(IdentifyPayload identify) {
     super.identify(identify);
 
-    String username = identify.traits().username();
-    KahunaAnalytics.setUsernameAndEmail(isNullOrEmpty(username) ? identify.userId() : username,
-        identify.traits().email());
-
-    KahunaAnalytics.setUserCredential(USERNAME_KEY, identify.traits().username());
-    KahunaAnalytics.setUserCredential(EMAIL_KEY, identify.traits().email());
-    KahunaAnalytics.setUserCredential(FACEBOOK_KEY, identify.traits().getString("facebook"));
-    KahunaAnalytics.setUserCredential(TWITTER_KEY, identify.traits().getString("twitter"));
-    KahunaAnalytics.setUserCredential(LINKEDIN_KEY, identify.traits().getString("linkedin"));
-
-    KahunaAnalytics.setUserAttributes(identify.traits().toStringMap());
+    Traits traits = identify.traits();
+    Map<String, String> userAttributes = new HashMap<>();
+    for (String key : traits.keySet()) {
+      if (KAHUNA_CREDENTIALS.contains(key)) {
+        // Only set credentials if it is a key recognized by Kahuna
+        KahunaAnalytics.setUserCredential(key, traits.getString(key));
+      } else {
+        // Set it as a user attribute otherwise
+        Object value = traits.get(key);
+        if (value instanceof Date) {
+          userAttributes.put(key, Utils.toISO8601Date((Date) value));
+        } else {
+          userAttributes.put(key, String.valueOf(value));
+        }
+      }
+    }
+    KahunaAnalytics.setUserAttributes(userAttributes);
   }
 
   @Override public void track(TrackPayload track) {
     super.track(track);
 
-    // Although not documented, Kahuna wants revenue in cents
-    KahunaAnalytics.trackEvent(track.event(), track.properties().getInt("quantity", 0),
-        (int) (track.properties().revenue() * 100));
+    String event = track.event();
+    int quantity = track.properties().getInt("quantity", -1);
+    double revenue = track.properties().revenue();
+    if (quantity == -1 && revenue == 0) {
+      // Kahuna requires revenue in cents
+      KahunaAnalytics.trackEvent(event);
+    } else {
+      KahunaAnalytics.trackEvent(event, quantity, (int) (revenue * 100));
+    }
+
+    if (VIEWED_PRODUCT_CATEGORY.equalsIgnoreCase(event)) {
+      trackViewedProductCategory(track);
+    } else if (VIEWED_PRODUCT.equalsIgnoreCase(event)) {
+      trackViewedProduct(track);
+      trackViewedProductCategory(track);
+    } else if (ADDED_PRODUCT.equalsIgnoreCase(event)) {
+      trackAddedProduct(track);
+      trackAddedProductCategory(track);
+    } else if (COMPLETED_ORDER.equalsIgnoreCase(event)) {
+      trackCompletedOrder(track);
+    }
```

Comment:
Can we move this logic up above track event? We prefer tracking the attributes before tracking the event.